File: easy/110-balanced-binary-tree/110-balanced-binary-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Boolean}
def is_balanced(root)
    dfs(root)[0]
end

def dfs(node)
    return [true, 0] if !node
    
    left = dfs(node.left)
    right = dfs(node.right)
    balanced = left[0] && right[0] && (left[1] - right[1]).abs < 2
    
    [balanced, 1 + [left[1], right[1]].max]
end

File: easy/110-balanced-binary-tree/110-balanced-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */


const isBalanced = function(root) {
  const dfs = node => {
      if (!node) return [true, 0];
      
      const left = dfs(node.left),
            right = dfs(node.right),
            balanced = left[0] && right[0] &&
                       Math.abs(left[1] - right[1]) < 2;
      
      return [balanced, 1 + Math.max(left[1], right[1])];
  }
  
  return dfs(root)[0];
}